storage: persist fully nontrival ReplicaState

tbg opened the pull request:
Previously, there was some ambiguity about what parts of a Range were
actually written. For example, a nil lease and a zero-value lease were
both allowed and treated equally (in fact, Lease was the only field lacking
the treatment).

Proposer-evaluated KV will use the same container, ReplicaState, to transmit
changes to the state. This makes it worthwhile to be somewhat stricter here.
For example, FrozenStatus is now an enum whose zero value we never want to see
persisted to disk, and consequently we don't ever want to synthesize it from
disk. We don't (or at least shouldn't), but now we assert similarly for all
fields.

There are mild migration concerns here for clusters which are running with
#9883, but that PR will not be part of a release until 2016-10-20.

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9970)

<!-- Reviewable:end -->

tamird commented:
> We don't (or at least shouldn't), but now we assert similarly for all
> fields.

Something off about this sentence.

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_command.go, line 2722 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PU9cja4n9kaJbXjv:-KU-PU9cja4n9kaJbXjw:b-jalcz) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2722)):*
> ```Go
> 		// This looks like an innocuous performance improvement, but it's more than
> 		// that - it ensures that we properly initialize the timestamp cache, which
> 		// is only populated on the lease holder, from that of the original Range.
> ```

s/Range/Replica/, i think

---

*[pkg/storage/replica_command.go, line 2732 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PgTZS-5OjejPhQQJ:-KU-PgTZS-5OjejPhQQK:b-8ax0yo) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2732)):*
> ```Go
> 		//   Replica. However, only node one's timestamp cache contains information
> 		//   about the read of key 'd' in the first place.
> 		// - node two becomes the lease holder for [c,e). Its timestamp cache does
> ```

does *not*?

---

*[pkg/storage/replica_command.go, line 2736 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PltNZU-3Ppv0X5a8:-KU-PltNZU-3Ppv0X5a9:bfnz0fd) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2736)):*
> ```Go
> 		// - node two can illegally propose a write to 'd' at a lower timestamp.
> 		//
> 		// TODO(tschottdorf): why would this use r.store.Engine() an not the
> ```

s/an /and /

---

*[pkg/storage/replica_command.go, line 2743 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-VRhNOsr0pdr1k_M0:-KU-VRhNOsr0pdr1k_M1:b2ni46c) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2743)):*
> ```Go
> 		}
> 		if (*leftLease == roachpb.Lease{}) {
> 			panic(leftLease)
> ```

hm.

---

*[pkg/storage/replica_state.go, line 498 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-P8h8K8B-yCFHF80l:-KU-P8h8K8B-yCFHF80m:b80vaqq) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_state.go#L498)):*
> ```Go
> 	if lease, err := loadLease(ctx, eng, desc.RangeID); err != nil {
> 		return enginepb.MVCCStats{}, errors.Wrap(err, "error reading lease")
> 	} else if (lease != nil && *lease != roachpb.Lease{}) {
> ```

so both are possible?

---

*[pkg/storage/store_test.go, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-OYKGI4h98-hDT61s:-KU-OYKGI4h98-hDT61t:b33i3x4) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/store_test.go#L152)):*
> ```Go
> 	defer func() {
> 		if !ranToCompletion {
> 			stopper.Stop()
> ```

this is pretty awful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-:-KU-OEO87jf4FaIuje9-:b-pkfg58)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
LGTM

What are the migration concerns? #9883 may not be in a release yet, but we're running clusters internally with non-released binaries. 

---

Reviewed 7 of 7 files at r1.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/storage/replica_state.go, line 233 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU00goAJEHvMnrQhrJO:-KU00goAJEHvMnrQhrJP:bkmi0x5) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_state.go#L233)):*
> ```Go
> ) error {
> 	if (truncState == roachpb.RaftTruncatedState{}) {
> 		return errors.New("cannot persist trivial RaftTruncatedState")
> ```

s/trivial/empty/

---

*[pkg/storage/replica_state.go, line 496 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU018fw-hO1ZTTMbZql:-KU018fw-hO1ZTTMbZqm:b-3zieh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_state.go#L496)):*
> ```Go
> 	s.Lease = lease
> 
> 	if lease, err := loadLease(ctx, eng, desc.RangeID); err != nil {
> ```

The shadowing of `lease` is confusing. 

Why does `writeInitialState` need to read? What could be written before the "initial" state?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-:-KU01YaFjqBwFTKnPJm8:bphpbqs)*
<!-- Sent from Reviewable.io -->

tbg committed:
storage: persist fully nontrival ReplicaState

Previously, there was some ambiguity about what parts of a Range were
actually written. For example, a nil lease and a zero-value lease were
both allowed and treated equally (in fact, Lease was the only field lacking
the treatment).

Proposer-evaluated KV will use the same container, ReplicaState, to transmit
changes to the state. This makes it worthwhile to be somewhat stricter here.
For example, FrozenStatus is now an enum whose zero value we never want to see
persisted to disk, and consequently we don't ever want to synthesize it from
disk. We don't, and now we assert similarly for all fields.

There are mild migration concerns here for clusters which are running with
\#9883, but that PR will not be part of a release until 2016-10-20.

tschottdorf commented:
Now that I think about it, I'm not sure there are any. I was worried that nodes would error out trying to write a state with a `nil` Lease, but we never return a `nil` lease from `loadState`, so that wouldn't happen in practice (similarly, `loadState` never returns the `FROZEN_UNSPECIFIED` status). We should be good to go, though I've missed things in the past. In this case, we would always be able to just remove the assertions again and be ok, though.

---

Review status: 4 of 7 files reviewed at latest revision, 9 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 2722 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PU9cja4n9kaJbXjv:-KU1ynpJUcqE0qh-KrJV:bctketx) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2722)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/Range/Replica/, i think
</details>
I don't think so, we populate "from the lease holder of the original Range".

---

*[pkg/storage/replica_command.go, line 2732 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PgTZS-5OjejPhQQJ:-KU1yrum6esdfEstuONk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2732)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> does *not*?
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2736 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PltNZU-3Ppv0X5a8:-KU1yxU4BLJkw39KRZ5y:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2736)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/an /and /
</details>
Done.

---

*[pkg/storage/replica_command.go, line 2743 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-VRhNOsr0pdr1k_M0:-KU1yy7v_pK0O-FBjzBo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2743)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> hm.
</details>
Done.

---

*[pkg/storage/replica_state.go, line 233 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU00goAJEHvMnrQhrJO:-KU1yz47tLHXnpIUlYsj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_state.go#L233)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/trivial/empty/
</details>
Done.

---

*[pkg/storage/replica_state.go, line 496 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU018fw-hO1ZTTMbZql:-KU1z1xnbnXI4X_7ToLl:bfspwbo) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_state.go#L496)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> The shadowing of `lease` is confusing. 
> 
> Why does `writeInitialState` need to read? What could be written before the "initial" state?
</details>
Done.

It reads the lease only for the assertion (i.e. makes sure we're not clobbering an existing range. This could happen if we screwed up the canApplySnapshot checks).

---

*[pkg/storage/replica_state.go, line 498 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-P8h8K8B-yCFHF80l:-KU1zC3w9KXLov7vfYZJ:b-b0e6of) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_state.go#L498)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> so both are possible?
</details>
No, `loadLease` always returns non-nil, but I don't want to burden readers with that knowledge.

---

*[pkg/storage/store_test.go, line 152 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-OYKGI4h98-hDT61s:-KU1zN2D70ZCstOBFEz2:boeid) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/store_test.go#L152)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this is pretty awful.
</details>
I agree, and I thought this was the best we could do, but actually... `t.Failed()` to the rescue!

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-:-KU1xvBWdcAncXp1FkGz:bc9g2f1)*
<!-- Sent from Reviewable.io -->

petermattis commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 4 of 7 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 2736 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PltNZU-3Ppv0X5a8:-KU2ZN1Io4UxEt7nVA1r:b-u5sggf) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2736)):*
<details><summary><i>Previously, tschottdorf (Tobias Schottdorf) wrote…</i></summary>
> Done.
</details>
Perhaps for the same reason as the comment above `loadHardState` indicates, though that doesn't seem necessary as we had `Replica.raftMu` for the left-hand side locked before we created the batch. I'd leave as-is, though, unless you're worried that something else in the batch could have written the lease.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-:-KU2_6MTMLR6KPXWHVE3:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tschottdorf commented:
Review status: 4 of 7 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[pkg/storage/replica_command.go, line 2736 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970#-KU-PltNZU-3Ppv0X5a8:-KU2ck8ddlyLiTZx4c--:b-am5n40) ([raw file](https://github.com/cockroachdb/cockroach/blob/498a8797bc03f5cb89157bbc3bfcd7e2e0a26833/pkg/storage/replica_command.go#L2736)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary>
> Perhaps for the same reason as the comment above `loadHardState` indicates, though that doesn't seem necessary as we had `Replica.raftMu` for the left-hand side locked before we created the batch. I'd leave as-is, though, unless you're worried that something else in the batch could have written the lease.
</details>
There's one obvious way in which this could backfire - we're currently sending out a new request for the lease and wait for that, but the obvious smart thing to do would be to prepend the lease request and then send the batch (cutting latency in half), in which case there *would* be another lease in the batch...

Also, HardState is different because that's "outside of the state machine". The lease is only written as application of a Raft command, so there's no way someone else is writing it concurrently.

Won't poke this now, but I think the confused comment is legit and should be addressed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9970)*
<!-- Sent from Reviewable.io -->
